File: android/cenario/src/main/kotlin/dev/odeon/android/ui/locadora/Midia.kt
package dev.odeon.android.ui.locadora

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import dev.odeon.android.ui.Texto
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.sp
import coil3.compose.AsyncImage
import dev.odeon.android.ui.Cores
import dev.odeon.android.ui.Serifada
import kotlin.math.cos
import kotlin.math.sin

/// O disco e a fita — a mídia que sai de dentro da caixa.
///
/// ## Eles são **desenhados**, e não imagens
///
/// É a régua de «zero bytes» do §15, a mesma que rendeu o avatar por hash e a
/// marquise: um disco é um círculo com anéis e um reflexo, e uma fita é um
/// retângulo com dois carretéis. Desenhados, eles escalam em qualquer tamanho,
/// giram sem custo e não precisam de arte que alguém tenha que produzir pra
/// 17.930 obras.
///
/// E, ao contrário do pôster, eles são **iguais pra todo mundo** — um DVD é um
/// DVD. O que muda de obra pra obra é a cor que o disco reflete, e ela sai da
/// `dominant_color` que o servidor já extraiu.

/// O disco, visto de frente e inclinado pela pose.
///
/// ## A inclinação é o que faz ele parecer disco
///
/// Um círculo chapado é uma moeda de desenho animado. O que o olho reconhece
/// como disco é a **elipse** — o círculo visto de esguelha — mais o
/// brilho que corre pela superfície quando ele gira. Aqui a elipse vem de
/// achatar a altura pelo cosseno da inclinação, que é a projeção de um círculo
/// num plano girado, sem precisar de matriz nenhuma.
@Composable
fun Disco(
    tamanho: Dp,
    cor: Color,
    pose: Pose,
    modifier: Modifier = Modifier,
    /// Ele gira sozinho? Só quando está tocando — um disco parado no ar é um
    /// objeto; um disco girando é uma máquina em funcionamento.
    girando: Boolean = false,
    /// ## A arte do rótulo — «tenta colocar o CD real»
    ///
    /// A queixa do dono era exata: o disco prateado genérico é um **DVD-R
    /// queimado em casa**, e uma locadora não aluga DVD-R. Disco prensado tem a
    /// arte da obra impressa de furo a borda — é a foto dos discos da Disney que
    /// ele mandou de referência.
    ///
    /// A arte é a mesma capa que a caixa já carrega, e `null` degrada pro
    /// prateado de antes (§24): metade do acervo não tem pôster, e nesses o
    /// disco sem impressão é a verdade.
    arte: String? = null,
) {
    /// ⚠️ **A animação só existe quando o disco gira**, e não é economia
    /// teórica.
    ///
    /// A primeira versão criava a `rememberInfiniteTransition` sempre e só usava
    /// o valor quando `girando` — ou seja, o Compose redesenhava a tela 60 vezes
    /// por segundo pra pintar um disco parado. O sintoma apareceu no emulador de
    /// um jeito inesperado: o `screencap` passou a devolver quadros pretos, que é
    /// o que uma superfície em animação contínua faz numa captura.
    ///
    /// Num aparelho de verdade o sintoma seria outro e pior: bateria indo embora
    /// com a caixa aberta na tela.
    val volta = if (!girando) {
        0f
    } else {
        val transicao = rememberInfiniteTransition(label = "disco girando")
        val animado by transicao.animateFloat(
            initialValue = 0f,
            targetValue = 360f,
            animationSpec = infiniteRepeatable(
                animation = tween(durationMillis = 3200, easing = LinearEasing),
                repeatMode = RepeatMode.Restart,
            ),
            label = "volta do disco",
        )
        animado
    }

    /// O achatamento: cosseno da inclinação vertical, com um piso pra o
    /// disco nunca virar uma linha — de perfil ele sumiria, e sumir no
    /// meio de um gesto parece defeito.
    val achatado = cos(Math.toRadians(pose.giroX.toDouble())).toFloat().coerceIn(0.35f, 1f)
    val giroNaTela = pose.giroY * 0.35f + volta

    /// ⚠️ O achatamento saiu do desenho e virou `graphicsLayer` do conjunto:
    /// com a arte impressa, achatar oval por oval deixaria a imagem redonda
    /// sobre anéis achatados — a elipse tem que ser do **disco inteiro**.
    Box(
        modifier
            .size(tamanho)
            .graphicsLayer { scaleY = achatado },
    ) {
        if (arte != null) {
            AsyncImage(
                model = arte,
                contentDescription = null,
                contentScale = ContentScale.Crop,
                modifier = Modifier
                    .fillMaxSize(0.96f)
                    .align(Alignment.Center)
                    .clip(CircleShape)
                    /// A arte gira **com o disco**; o brilho, não — reflexo é da
                    /// luz da sala, e a luz não roda junto com o rótulo.
                    .graphicsLayer { rotationZ = giroNaTela },
            )
        }
        Canvas(Modifier.fillMaxSize()) {
            val raio = size.minDimension / 2f
            val centro = Offset(size.width / 2f, size.height / 2f)

            rotate(degrees = giroNaTela, pivot = centro) {
                desenharDisco(centro, raio, cor, temArte = arte != null)
            }
            /// O verniz: a luz varrendo o policarbonato por cima da impressão.
            /// Fora do `rotate` — ver o comentário da arte.
            desenharVerniz(centro, raio)
        }
    }
}

/// O desenho em si, separado pra caber num `Canvas` de qualquer lugar.
///
/// Com arte, o corpo não é pintado — a impressão está por baixo e o que o
/// desenho põe são as partes que **nenhuma impressão cobre** num disco de
/// verdade: a borda transparente, o espelho do cubo, o furo. Sem arte, o corpo
/// prateado de antes continua sendo a verdade (§24).
private fun DrawScope.desenharDisco(centro: Offset, raio: Float, cor: Color, temArte: Boolean) {
    if (!temArte) {
        /// O corpo: prateado com um leve tom da obra. A cor da obra entra **na
        /// reflexão**, não no plástico — §15, «a cor da obra só toca arte».
        drawCircle(
            brush = Brush.linearGradient(
                0.0f to Color(0xFFB9BDC7),
                0.35f to Color(0xFFEDEFF3),
                0.5f to cor.copy(alpha = 0.55f),
                0.65f to Color(0xFFEDEFF3),
                1.0f to Color(0xFF8E939D),
            ),
            radius = raio,
            center = centro,
        )
    }

    /// A borda de leitura: o anel fino de policarbonato sem impressão, com o
    /// vinco escuro onde os dados acabam. É o que separa «círculo com imagem»
    /// de «disco impresso» — todo prensado tem esse aro nu.
    drawCircle(
        color = Color(0xFFC9CDD6).copy(alpha = if (temArte) 0.9f else 0.5f),
        radius = raio * 0.98f,
        center = centro,
        style = Stroke(width = raio * 0.05f),
    )
    drawCircle(
        color = Color.Black.copy(alpha = 0.22f),
        radius = raio * 0.945f,
        center = centro,
        style = Stroke(width = raio * 0.018f),
    )

    /// O cubo: o anel-espelho em volta do furo — a parte prateada que sobra em
    /// qualquer rótulo —, o plástico translúcido e o furo em si.
    drawCircle(
        brush = Brush.sweepGradient(
            listOf(
                Color(0xFFD8DCE4),
                Color(0xFFAEB3BE),
                Color(0xFFEDEFF3),
                Color(0xFFB9BDC7),
                Color(0xFFD8DCE4),
            ),
            center = centro,
        ),
        radius = raio * 0.34f,
        center = centro,
    )
    drawCircle(color = Color(0xFFE4E7ED), radius = raio * 0.22f, center = centro)
    drawCircle(color = Cores.fundoAfundado, radius = raio * 0.13f, center = centro)
    drawCircle(
        color = Color.White.copy(alpha = 0.35f),
        radius = raio * 0.135f,
        center = centro,
        style = Stroke(width = raio * 0.015f),
    )
}

/// A luz sobre o policarbonato: o facho diagonal e o arco-íris da difração.
///
/// Separados do disco porque **não giram com ele** — reflexo é da lâmpada da
/// sala, e a lâmpada fica parada enquanto o rótulo roda.
private fun DrawScope.desenharVerniz(centro: Offset, raio: Float) {
    /// O facho: um quarto de volta de brilho branco, na diagonal de cima —
    /// a mesma lâmpada do `luzNoLado`, batendo no plástico.
    drawArc(
        brush = Brush.sweepGradient(
            listOf(
                Color.Transparent,
                Color.White.copy(alpha = 0.30f),
                Color.White.copy(alpha = 0.05f),
                Color.Transparent,
            ),
            center = centro,
        ),
        startAngle = 205f,
        sweepAngle = 70f,
        useCenter = false,
        topLeft = Offset(centro.x - raio * 0.66f, centro.y - raio * 0.66f),
        size = Size(raio * 1.32f, raio * 1.32f),
        style = Stroke(width = raio * 0.6f),
    )

    /// O arco-íris do policarbonato: duas faixas curtas, opostas, bem
    /// transparentes. É o detalhe que faz o olho dizer «disco» antes de ler
    /// qualquer coisa.
    listOf(20f, 200f).forEach { comeco ->
        drawArc(
            brush = Brush.sweepGradient(
                listOf(
                    Color.Transparent,
                    Color(0x4433CCFF),
                    Color(0x44FF66AA),
                    Color(0x4466FF99),
                    Color.Transparent,
                ),
                center = centro,
            ),
            startAngle = comeco,
            sweepAngle = 52f,
            useCenter = false,
            topLeft = Offset(centro.x - raio * 0.8f, centro.y - raio * 0.8f),
            size = Size(raio * 1.6f, raio * 1.6f),
            style = Stroke(width = raio * 0.36f),
        )
    }
}

/// A fita VHS: a carcaça, a janela e os dois carretéis.
///
/// ## O carretel é o mostrador, e é isso que a fita tem de especial
///
/// Num DVD não dá pra ver onde o filme parou. Numa fita **dá**: o rolo da
/// esquerda engorda conforme ela roda, e o da direita emagrece. É a mesma
/// informação de uma barra de progresso, só que dita pelo objeto — e é o que faz
/// «rebobinar» ser uma coisa que se entende sem explicação.
@Composable
fun FitaVHS(
    largura: Dp,
    cor: Color,
    modifier: Modifier = Modifier,
    /// Quanto da fita já rodou, de 0 a 1. É o que decide a grossura dos rolos.
    andado: Float = 0f,
    /// O quanto os carretéis já giraram, em voltas. Quem anima é quem chama —
    /// aqui é só desenho, e é o que deixa a tela de rebobinar controlar a
    /// velocidade do jeito dela.
    voltas: Float = 0f,
    /// ## O título no rótulo — «a original tinha detalhes do filme»
    ///
    /// A referência do dono é a fita d'*O Rei Leão*: o rótulo de papel dela é
    /// **impresso**, com o nome do filme — não um retângulo bege mudo. `null`
    /// deixa o papel em branco, que é o que uma fita sem identificação tem.
    titulo: String? = null,
) {
    val alturaDaFita = largura * 0.58f
    Box(modifier.size(largura, alturaDaFita)) {
        Canvas(Modifier.fillMaxSize()) {
            /// ⚠️ Os rolos moram **dentro da janela**, e a foto cobrou: com o
            /// centro no meio da carcaça (0,5) e raio de 0,30, o rolo cheio
            /// vazava o vidro e invadia o rótulo. O centro é o centro da janela
            /// (0,46) e o raio máximo é o que cabe nela com 1% de folga.
            val r = size.height * 0.27f
            val esquerda = Offset(size.width * 0.31f, size.height * 0.46f)
            val direita = Offset(size.width * 0.69f, size.height * 0.46f)

            /// A carcaça preta, com o topo levemente mais claro — a fita é
            /// plástico fosco, e plástico fosco pega luz de cima.
            drawRoundRect(
                brush = Brush.verticalGradient(
                    listOf(Color(0xFF2A2A2E), Color(0xFF141416)),
                ),
                cornerRadius = androidx.compose.ui.geometry.CornerRadius(size.height * 0.06f),
            )

            /// Os frisos de pega — as linhas em relevo que toda carcaça tem nas
            /// beiradas. Detalhe barato que tira o «retângulo liso» da fita.
            listOf(0.045f, 0.075f, 0.925f, 0.955f).forEach { x ->
                drawLine(
                    color = Color.White.copy(alpha = 0.05f),
                    start = Offset(size.width * x, size.height * 0.12f),
                    end = Offset(size.width * x, size.height * 0.88f),
                    strokeWidth = size.width * 0.008f,
                )
            }

            /// A janela: o retângulo transparente por onde se veem os rolos —
            /// com a moldura clara do acrílico, que a primeira versão não tinha.
            drawRoundRect(
                color = Color(0xFF0B0B0D),
                topLeft = Offset(size.width * 0.12f, size.height * 0.16f),
                size = Size(size.width * 0.76f, size.height * 0.60f),
                cornerRadius = androidx.compose.ui.geometry.CornerRadius(size.height * 0.04f),
            )
            drawRoundRect(
                color = Color.White.copy(alpha = 0.10f),
                topLeft = Offset(size.width * 0.12f, size.height * 0.16f),
                size = Size(size.width * 0.76f, size.height * 0.60f),
                cornerRadius = androidx.compose.ui.geometry.CornerRadius(size.height * 0.04f),
                style = Stroke(width = size.height * 0.012f),
            )

            /// Os dois rolos. O da esquerda cresce com o que já rodou; o da
            /// direita é o complemento — a fita tem comprimento fixo, e o que
            /// sai de um lado entra no outro.
            desenharCarretel(esquerda, r, 0.34f + 0.62f * andado, voltas, cor)
            desenharCarretel(direita, r, 0.34f + 0.62f * (1f - andado), -voltas, cor)

            /// Os quatro parafusos dos cantos — a carcaça é aparafusada, e são
            /// eles que dizem «objeto montado» em vez de «desenho de fita».
            listOf(
                Offset(size.width * 0.035f, size.height * 0.07f),
                Offset(size.width * 0.965f, size.height * 0.07f),
                Offset(size.width * 0.035f, size.height * 0.93f),
                Offset(size.width * 0.965f, size.height * 0.93f),
            ).forEach { onde ->
                drawCircle(Color(0xFF0A0A0C), radius = size.width * 0.012f, center = onde)
                drawCircle(
                    Color.White.copy(alpha = 0.18f),
                    radius = size.width * 0.012f,
                    center = onde,
                    style = Stroke(width = size.width * 0.004f),
                )
            }

            /// A etiqueta de papel, na parte de baixo, com a **cinta** na cor da
            /// obra — as fitas impressas levavam uma faixa colorida no rótulo, e
            /// a cor é a `dominant_color` que o servidor já extraiu.
            drawRoundRect(
                color = Color(0xFFE8E2D2).copy(alpha = 0.92f),
                topLeft = Offset(size.width * 0.16f, size.height * 0.79f),
                size = Size(size.width * 0.68f, size.height * 0.16f),
                cornerRadius = androidx.compose.ui.geometry.CornerRadius(2f),
            )
            drawRect(
                color = cor.copy(alpha = 0.85f),
                topLeft = Offset(size.width * 0.16f, size.height * 0.79f),
                size = Size(size.width * 0.68f, size.height * 0.025f),
            )
        }

        /// O título, impresso no papel. É `Text` e não desenho porque é texto de
        /// verdade — elipsado, medido, no corpo que a largura da fita permitir.
        if (titulo != null) {
            Texto(
                text = titulo,
                style = TextStyle(
                    fontFamily = Serifada,
                    fontSize = (largura.value * 0.032f).sp,
                    letterSpacing = 0.06.em,
                    color = Color(0xFF3A3428),
                ),
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                textAlign = TextAlign.Center,
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .fillMaxWidth(0.62f)
                    .padding(bottom = alturaDaFita * 0.048f),
            )
        }
    }
}

/// Um carretel: o rolo de fita enrolado e os dentes do eixo.
private fun DrawScope.desenharCarretel(
    centro: Offset,
    raioMaximo: Float,
    quanto: Float,
    voltas: Float,
    cor: Color,
) {
    val raio = raioMaximo * quanto.coerceIn(0.2f, 1f)

    /// O rolo de fita: marrom escuro, com um brilho fraco da cor da obra em
    /// cima — é a luz da sala batendo no filme, e é o que impede o rolo de ser
    /// um disco preto.
    drawCircle(
        brush = Brush.radialGradient(
            listOf(Color(0xFF3A2A22), Color(0xFF1E1512)),
            center = centro,
            radius = raio,
        ),
        radius = raio,
        center = centro,
    )
    drawCircle(
        color = cor.copy(alpha = 0.10f),
        radius = raio,
        center = centro,
    )

    /// O eixo branco e os dentes, que são o que **mostra o giro**: sem eles, um
    /// rolo redondo girando é um rolo redondo parado.
    drawCircle(color = Color(0xFFD5D5D8), radius = raioMaximo * 0.28f, center = centro)
    val angulo = voltas * 360f
    repeat(6) { i ->
        val a = Math.toRadians((angulo + i * 60f).toDouble())
        val ponta = Offset(
            centro.x + (raioMaximo * 0.26f * cos(a)).toFloat(),
            centro.y + (raioMaximo * 0.26f * sin(a)).toFloat(),
        )
        drawLine(
            color = Color(0xFF6E6E74),
            start = centro,
            end = ponta,
            strokeWidth = raioMaximo * 0.09f,
        )
    }
    drawCircle(color = Color(0xFF17171A), radius = raioMaximo * 0.10f, center = centro)
}

File: ios/Odeon/Interface/Midia.swift
import SwiftUI

/// O disco e a fita — a mídia que sai de dentro da caixa.
///
/// ## ⚠️ Eles são **desenhados**, e não imagens
///
/// É a régua de «zero bytes» da casa, a mesma que rendeu a marquise e a
/// arandela: um disco é um círculo com anéis e um reflexo; uma fita é um
/// retângulo com dois carretéis. Desenhados, eles escalam em qualquer tamanho e
/// não precisam de arte que alguém tenha de produzir pra 17.930 obras.
///
/// E, ao contrário do pôster, eles são **iguais pra todo mundo** — um DVD é um
/// DVD. O que muda de obra pra obra é a cor que o disco reflete, e ela sai da
/// `dominant_color` que o servidor já extraiu.
enum Midia {}

/// O disco, visto de frente e inclinado pela pose.
///
/// ## A inclinação é o que faz ele parecer disco
///
/// Um círculo chapado é uma moeda de desenho animado. O que o olho reconhece
/// como disco é a **elipse** — o círculo visto de esguelha — mais o brilho que
/// corre pela superfície. A elipse vem de achatar a altura pelo cosseno da
/// inclinação, que é a projeção de um círculo num plano girado.
struct Disco: View {
    let tamanho: CGFloat
    let cor: Color
    let pose: Pose
    /// A arte do rótulo. ⚠️ **Disco prensado tem a arte impressa de furo a
    /// borda** — o prateado genérico é um DVD-R queimado em casa, e uma locadora
    /// não aluga DVD-R. `nil` degrada pro prateado, que é a verdade nos 48% do
    /// acervo sem pôster (§24).
    let capa: String?
    let odeon: RepositorioOdeon

    /// ⚠️ O achatamento tem **piso**: de perfil o disco viraria uma linha e
    /// sumiria no meio do gesto, o que parece defeito.
    private var achatado: CGFloat { min(max(cos(pose.giroX * .pi / 180), 0.35), 1) }

    /// ⚠️ Ele acompanha o giro da caixa **por um terço**: está solto no estojo,
    /// não parafusado na tampa.
    private var giroNaTela: CGFloat { pose.giroY * 0.35 }

    var body: some View {
        ZStack {
            /// ⚠️ A arte gira **com o disco**; o verniz, não — reflexo é da luz da
            /// sala, e a luz não roda junto com o rótulo.
            if capa != nil {
                ArteDoOdeon(odeon: odeon, caminho: capa)
                    .frame(width: tamanho * 0.96, height: tamanho * 0.96)
                    .clipShape(.circle)
                    .rotationEffect(.degrees(giroNaTela))
            }

            Canvas { contexto, size in
                let centro = CGPoint(x: size.width / 2, y: size.height / 2)
                let raio = min(size.width, size.height) / 2
                desenhar(contexto, centro: centro, raio: raio, temArte: capa != nil)
                verniz(contexto, centro: centro, raio: raio)
            }
        }
        .frame(width: tamanho, height: tamanho)
        /// ⚠️ O achatamento é do **conjunto**, e não oval por oval: achatando cada
        /// anel, a arte sairia redonda sobre anéis achatados. A elipse tem de ser
        /// do disco inteiro.
        .scaleEffect(x: 1, y: achatado)
    }

    private func desenhar(_ contexto: GraphicsContext, centro: CGPoint,
                          raio: CGFloat, temArte: Bool) {
        func circulo(_ r: CGFloat) -> Path {
            Path(ellipseIn: CGRect(x: centro.x - r, y: centro.y - r, width: r * 2, height: r * 2))
        }

        if !temArte {
            /// ⚠️ A cor da obra entra **na reflexão**, não no plástico: «a cor da
            /// obra só toca arte».
            contexto.fill(circulo(raio), with: .linearGradient(
                Gradient(stops: [
                    .init(color: Color(hex: 0xB9BDC7), location: 0),
                    .init(color: Color(hex: 0xEDEFF3), location: 0.35),
                    .init(color: cor.opacity(0.55), location: 0.5),
                    .init(color: Color(hex: 0xEDEFF3), location: 0.65),
                    .init(color: Color(hex: 0x8E939D), location: 1),
                ]),
                startPoint: CGPoint(x: centro.x - raio, y: centro.y - raio),
                endPoint: CGPoint(x: centro.x + raio, y: centro.y + raio),
            ))
        }

        /// A borda de leitura: o aro de policarbonato **sem impressão**, com o
        /// vinco escuro onde os dados acabam. É o que separa «círculo com imagem»
        /// de «disco prensado» — todo prensado tem esse aro nu.
        contexto.stroke(circulo(raio * 0.98),
                        with: .color(Color(hex: 0xC9CDD6).opacity(temArte ? 0.9 : 0.5)),
                        lineWidth: raio * 0.05)
        contexto.stroke(circulo(raio * 0.945),
                        with: .color(.black.opacity(0.22)), lineWidth: raio * 0.018)

        /// O cubo: o anel-espelho em volta do furo — a parte prateada que sobra em
        /// qualquer rótulo —, o plástico translúcido e o furo.
        contexto.fill(circulo(raio * 0.34), with: .conicGradient(
            Gradient(colors: [
                Color(hex: 0xD8DCE4), Color(hex: 0xAEB3BE), Color(hex: 0xEDEFF3),
                Color(hex: 0xB9BDC7), Color(hex: 0xD8DCE4),
            ]), center: centro, angle: .zero,
        ))
        contexto.fill(circulo(raio * 0.22), with: .color(Color(hex: 0xE4E7ED)))
        contexto.fill(circulo(raio * 0.13), with: .color(Cores.fundo))
        contexto.stroke(circulo(raio * 0.135),
                        with: .color(.white.opacity(0.35)), lineWidth: raio * 0.015)
    }

    /// A luz sobre o policarbonato.
    ///
    /// ⚠️ Separada do disco porque **não gira com ele**: reflexo é da lâmpada da
    /// sala, e a lâmpada fica parada enquanto o rótulo roda.
    private func verniz(_ contexto: GraphicsContext, centro: CGPoint, raio: CGFloat) {
        contexto.fill(
            Path(ellipseIn: CGRect(x: centro.x - raio, y: centro.y - raio,
                                   width: raio * 2, height: raio * 2)),
            with: .conicGradient(
                Gradient(stops: [
                    .init(color: .clear, location: 0),
                    .init(color: .white.opacity(0.30), location: 0.12),
                    .init(color: .white.opacity(0.05), location: 0.22),
                    .init(color: .clear, location: 0.34),
                    .init(color: .clear, location: 1),
                ]), center: centro, angle: .degrees(-40),
            ),
        )
    }
}

/// A fita VHS.
struct FitaVHS: View {
    let largura: CGFloat
    let titulo: String?

    var body: some View {
        Canvas { contexto, size in
            let r = size.height * 0.27
            let esquerda = CGPoint(x: size.width * 0.31, y: size.height * 0.46)
            let direita = CGPoint(x: size.width * 0.69, y: size.height * 0.46)

            /// A carcaça, com o topo mais claro — plástico fosco pega luz de cima.
            contexto.fill(
                Path(roundedRect: CGRect(origin: .zero, size: size),
                     cornerRadius: size.height * 0.06),
                with: .linearGradient(
                    Gradient(colors: [Color(hex: 0x2A2A2E), Color(hex: 0x141416)]),
                    startPoint: .zero, endPoint: CGPoint(x: 0, y: size.height),
                ),
            )

            /// Os frisos de pega — as linhas em relevo que toda carcaça tem nas
            /// beiradas. Detalhe barato que tira o «retângulo liso» da fita.
            for x in [0.045, 0.075, 0.925, 0.955] as [CGFloat] {
                var friso = Path()
                friso.move(to: CGPoint(x: size.width * x, y: size.height * 0.12))
                friso.addLine(to: CGPoint(x: size.width * x, y: size.height * 0.88))
                contexto.stroke(friso, with: .color(.white.opacity(0.05)),
                                lineWidth: size.width * 0.008)
            }

            /// ⚠️ A janela, e os rolos moram **dentro dela**: com o centro no meio
            /// da carcaça o rolo cheio vazava o vidro e invadia o rótulo. O centro
            /// é o da janela (0,46).
            contexto.fill(
                Path(roundedRect: CGRect(x: size.width * 0.12, y: size.height * 0.16,
                                         width: size.width * 0.76, height: size.height * 0.60),
                     cornerRadius: size.height * 0.04),
                with: .color(Color(hex: 0x0B0B0D)),
            )

            /// Os dois carretéis, e a fita esticada entre eles.
            for centro in [esquerda, direita] {
                contexto.fill(
                    Path(ellipseIn: CGRect(x: centro.x - r, y: centro.y - r,
                                           width: r * 2, height: r * 2)),
                    with: .color(Color(hex: 0x1A1A1E)),
                )
                contexto.stroke(
                    Path(ellipseIn: CGRect(x: centro.x - r * 0.42, y: centro.y - r * 0.42,
                                           width: r * 0.84, height: r * 0.84)),
                    with: .color(.white.opacity(0.18)), lineWidth: r * 0.10,
                )
            }
            var fita = Path()
            fita.move(to: CGPoint(x: esquerda.x, y: esquerda.y - r))
            fita.addLine(to: CGPoint(x: direita.x, y: direita.y - r))
            contexto.stroke(fita, with: .color(Color(hex: 0x3A2E24)), lineWidth: r * 0.16)

            /// ⚠️ O rótulo é **impresso**, com o nome do filme — a referência é a
            /// fita d'*O Rei Leão*, não um retângulo bege mudo.
            contexto.fill(
                Path(roundedRect: CGRect(x: size.width * 0.10, y: size.height * 0.80,
                                         width: size.width * 0.80, height: size.height * 0.14),
                     cornerRadius: 2),
                with: .color(Cores.papel),
            )
            if let titulo {
                let texto = contexto.resolve(
                    Text(titulo)
                        .font(.system(size: max(5, size.height * 0.085), weight: .semibold))
                        .foregroundStyle(Cores.tintaDoBilhete),
                )
                contexto.draw(texto, at: CGPoint(x: size.width / 2, y: size.height * 0.87))
            }
        }
        .frame(width: largura, height: largura * 0.58)
    }
}
